File: floe/jobdeck/color.py
"""Colour assignment for a jobdeck view - the MDPView rule as measured.

Three modes:

  MODE_IDENTIFIER   flat view: one colour per `$` identifier, ascending.
  MODE_LAYER        chip view: colour by LY (a specific LY/DT can be pinned).
  MODE_CHIP         one colour per CHIP block in deck order; the selected
                    identifier k is spliced in after k-1 CHIPs.

MDPView rotates through TEN colours (JOBDECK_PALETTE - exactly the Tk
colour names blue, yellow, red, pink, orange, white, purple, cyan,
magenta, green) assigned by POSITION in an ordered target list; the
eleventh target gets the first colour again (measured 2026-09-07).
`color_order()` builds that list as strings ("$2", "CHIP:ID001", "LY123")
so it can be compared with the MDPView layer list by eye and stored in a
report. Chip-mode colours therefore depend on the selection; identifier
and layer colours do not. Several selected identifiers at once in chip
mode is an extension (MDPView expands one at a time): inferred, not
measured.

RESERVE_PALETTE (49 colours organised like Calibre's default layer
colours) is kept selectable ("palette": "reserve") because its use is
not known; it is not a guess at MDPView's colours.

Nothing here renders and nothing is KLayout-specific: the result is
`render key -> "#rrggbb"`; the renderer (M2+) turns it into layer styles.
"""

from __future__ import annotations

import json
from dataclasses import dataclass, field

# MDPView's two jobdeck views (manual terms, user 2026-09-09): the LEVEL
# view lists the mask levels - the `$n` entries, MTITLE n naming them -
# and the CHIP view lists the CHIP blocks. The `$n` number is the mask
# level number (Artwork's MEBES job deck syntax: `CHIP name,(1,PATTERN,
# AD=...)`, "1" = level), so what the port called "identifier" is the
# level; "identifier" stays accepted as an alias. The LY/DT view is an
# extra of ours (source layer/datatype), not a manual term.
MODE_LEVEL = "level"
MODE_IDENTIFIER = MODE_LEVEL
MODE_LAYER = "layer"
MODE_CHIP = "chip"
MODES = (MODE_LEVEL, MODE_LAYER, MODE_CHIP)
MODE_ALIASES = {"identifier": MODE_LEVEL, "id": MODE_LEVEL,
                "levels": MODE_LEVEL, "chips": MODE_CHIP,
                "layers": MODE_LAYER}


def normalize_mode(mode) -> str:
    """'identifier' -> 'level'; validates against MODES."""
    m = MODE_ALIASES.get(str(mode).lower(), str(mode).lower())
    if m not in MODES:
        raise ValueError("jobdeck view must be one of %s, got %r"
                         % (MODES, mode))
    return m

JOBDECK_PALETTE = [
    "#0000ff", "#ffff00", "#ff0000", "#ffc0cb", "#ffa500",
    "#ffffff", "#a020f0", "#00ffff", "#ff00ff", "#00ff00",
]
JOBDECK_NAMES = ["blue", "yellow", "red", "pink", "orange",
                 "white", "purple", "cyan", "magenta", "green"]

RESERVE_PALETTE = [
    "#ff0000", "#00ff00", "#2626ff", "#ffff00", "#ff00ff",
    "#00ffff", "#ff8000", "#80ff00", "#00ff80", "#0080ff",
    "#8000ff", "#ff0080", "#ff7373", "#73ff73", "#7373ff",
    "#ffff73", "#ff73ff", "#73ffff", "#ffb973", "#b9ff73",
    "#73ffb9", "#73b9ff", "#b973ff", "#ff73b9", "#b80000",
    "#00b800", "#4949b8", "#b8b800", "#b800b8", "#00b8b8",
    "#b85c00", "#5cb800", "#00b85c", "#005cb8", "#732eb8",
    "#b8005c", "#d94141", "#41d941", "#4141d9", "#d9d941",
    "#d941d9", "#41d9d9", "#d98d41", "#8dd941", "#41d98d",
    "#418dd9", "#8d41d9", "#d9418d", "#ffffff",
]

DEFAULT_PALETTE = JOBDECK_PALETTE
PALETTES = {"jobdeck": JOBDECK_PALETTE, "reserve": RESERVE_PALETTE}


def palette_color(palette, i: int) -> str:
    """Colour for position i; the palette rotates (wraps) like MDPView's."""
    if not palette:
        raise ValueError("empty palette")
    return palette[i % len(palette)]


def color_name(color: str) -> str:
    """'#0000ff' -> 'blue' for the ten jobdeck colours, else the hex."""
    c = str(color).lower()
    for hexval, name in zip(JOBDECK_PALETTE, JOBDECK_NAMES):
        if c == hexval:
            return name
    return color


def resolve_palette(spec) -> list[str]:
    if isinstance(spec, str):
        try:
            return list(PALETTES[spec.lower()])
        except KeyError:
            raise ValueError("unknown palette %r; use one of %s or a list "
                             "of colours" % (spec, sorted(PALETTES)))
    pal = [str(c) for c in spec]
    if not pal:
        raise ValueError("palette must not be empty")
    return pal


def entry_layer_pairs(entry, cross: bool = True) -> list:
    from .geom import entry_pairs
    return entry_pairs(entry, cross)


def key_identifier(idx) -> str:
    return "$%d" % int(idx)


def key_chip(chip_id) -> str:
    return "CHIP:%s" % chip_id


def key_layer(ly) -> str:
    return "LY%d" % int(ly)


def deck_chip_ids(deck) -> list[str]:
    """CHIP ids in deck order, deduplicated (a duplicate id is one target)."""
    out = []
    for c in deck.chips:
        if c.id not in out:
            out.append(c.id)
    return out


def chip_order(chip_ids, ids) -> list[str]:
    """CHIPs in the given order with each selected identifier k spliced in
    after k-1 CHIPs.

    >>> chip_order(["C1", "C2", "C3", "C4"], [2])
    ['CHIP:C1', '$2', 'CHIP:C2', 'CHIP:C3', 'CHIP:C4']
    >>> chip_order(["C1", "C2", "C3"], [1, 3])
    ['$1', 'CHIP:C1', 'CHIP:C2', '$3', 'CHIP:C3']

    An identifier larger than the CHIP count comes after the last CHIP.
    """
    out = []
    taken = 0
    for k in sorted({int(i) for i in ids}):
        want = min(max(k - 1, 0), len(chip_ids))
        while taken < want:
            out.append(key_chip(chip_ids[taken]))
            taken += 1
        out.append(key_identifier(k))
    out.extend(key_chip(c) for c in chip_ids[taken:])
    return out


def color_order(deck, mode: str, ids=None, cross: bool = True) -> list[str]:
    """The ordered colour targets for a mode: position i gets palette
    colour i mod the palette length."""
    if mode == MODE_IDENTIFIER:
        return [key_identifier(i) for i in deck.identifiers()]
    if mode == MODE_CHIP:
        sel = deck.identifiers() if ids is None else ids
        return chip_order(deck_chip_ids(deck), sel)
    if mode == MODE_LAYER:
        seen = set()
        for c in deck.chips:
            for e in c.entries:
                seen.update(ly for ly, _ in entry_layer_pairs(e, cross))
        return [key_layer(ly) for ly in sorted(seen)]
    raise ValueError("mode must be one of %s, got %r" % (MODES, mode))


def layer_pairs(deck, cross: bool = True) -> list:
    seen = set()
    for c in deck.chips:
        for e in c.entries:
            seen.update(entry_layer_pairs(e, cross))
    return sorted(seen)


@dataclass
class ColorScheme:
    mode: str = MODE_IDENTIFIER
    palette: list = field(default_factory=lambda: list(DEFAULT_PALETTE))
    # overrides keys: str(identifier) | str(ly) | "ly/dt" | chip id
    overrides: dict = field(default_factory=dict)
    cross_ly_dt: bool = True
    fallback: str = "#808080"

    def pin_layer(self, ly: int, color: str) -> None:
        self.overrides[str(ly)] = color

    def pin_layer_datatype(self, ly: int, dt: int, color: str) -> None:
        self.overrides["%d/%d" % (ly, dt)] = color

    def pin_identifier(self, idx: int, color: str) -> None:
        self.overrides[str(idx)] = color

    def pin_chip(self, chip_id: str, color: str) -> None:
        self.overrides[str(chip_id)] = color

    def order(self, deck, ids=None, mode=None) -> list[str]:
        return color_order(deck, mode or self.mode, ids, self.cross_ly_dt)

    def order_table(self, deck, ids=None, mode=None) -> list[dict]:
        """One row per position: pos, key, color, name, source (palette or
        pinned) - what a report stores and what is compared with the
        MDPView layer list."""
        mode = mode or self.mode
        rows = []
        for pos, key in enumerate(self.order(deck, ids, mode)):
            pinned = self._pinned(key)
            color = pinned or palette_color(self.palette, pos)
            rows.append({"pos": pos, "key": key, "color": color,
                         "name": color_name(color),
                         "source": "pinned" if pinned else "palette"})
        return rows

    def _pinned(self, key: str):
        if key.startswith("CHIP:"):
            return self.overrides.get(key[len("CHIP:"):])
        if key.startswith("$"):
            return self.overrides.get(key[1:])
        if key.startswith("LY"):
            return self.overrides.get(key[2:])
        return None

    def build(self, deck, ids=None) -> dict:
        """render key -> colour: {idx: c} / {chip_id: c} / {(ly, dt): c}."""
        order = self.order(deck, ids)
        pos = {key: i for i, key in enumerate(order)}

        def auto(key):
            return (palette_color(self.palette, pos[key]) if key in pos
                    else self.fallback)

        if self.mode == MODE_IDENTIFIER:
            return {i: self.overrides.get(str(i)) or auto(key_identifier(i))
                    for i in deck.identifiers()}
        if self.mode == MODE_CHIP:
            return {c: self.overrides.get(str(c)) or auto(key_chip(c))
                    for c in deck_chip_ids(deck)}
        out = {}
        for (ly, dt) in layer_pairs(deck, self.cross_ly_dt):
            out[(ly, dt)] = (self.overrides.get("%d/%d" % (ly, dt))
                             or self.overrides.get(str(ly))
                             or auto(key_layer(ly)))
        return out

    def color_of(self, colormap: dict, placement) -> str:
        """The colour one placement gets under this scheme."""
        if self.mode == MODE_IDENTIFIER:
            return colormap.get(placement.idx, self.fallback)
        if self.mode == MODE_CHIP:
            return colormap.get(placement.chip, self.fallback)
        return colormap.get((placement.ly, placement.dt), self.fallback)

    def save(self, path: str) -> None:
        with open(path, "w") as fh:
            json.dump({"mode": self.mode, "palette": self.palette,
                       "overrides": self.overrides,
                       "cross_ly_dt": self.cross_ly_dt,
                       "fallback": self.fallback}, fh, indent=2)

    @classmethod
    def load(cls, path: str) -> "ColorScheme":
        with open(path) as fh:
            d = json.load(fh)
        if "palette" in d:
            d["palette"] = resolve_palette(d["palette"])
        try:
            d["mode"] = normalize_mode(d.get("mode", MODE_LEVEL))
        except ValueError as exc:
            raise ValueError("colour file %s: %s" % (path, exc))
        return cls(**d)


def order_text(rows) -> str:
    """'CHIP:ID001 blue, $2 yellow, ...' from order_table rows."""
    return ", ".join("%s %s%s" % (r["key"], r["name"],
                                  " (pinned)" if r["source"] == "pinned"
                                  else "")
                     for r in rows)

File: rust/app-core/src/jobdeck/color.rs
//! Observed 10-color order. The analysis CLI's historical CHIP-block colors
//! are separate from the viewer's level -> source-chip tree.
use super::{
    geom::{entry_pairs, Placement},
    parser::JobDeck,
};
use crate::{catalog::regular_file, Error, Result};
use serde::{Deserialize, Serialize};
use serde_json::{json, Value};
use std::collections::{BTreeMap, BTreeSet};
use std::io::Read;
use std::path::Path;

pub const PALETTE: [&str; 10] = [
    "#0000ff", "#ffff00", "#ff0000", "#ffc0cb", "#ffa500", "#ffffff", "#a020f0", "#00ffff",
    "#ff00ff", "#00ff00",
];
const NAMES: [&str; 10] = [
    "blue", "yellow", "red", "pink", "orange", "white", "purple", "cyan", "magenta", "green",
];
const RESERVE: [&str; 49] = [
    "#ff0000", "#00ff00", "#2626ff", "#ffff00", "#ff00ff", "#00ffff", "#ff8000", "#80ff00",
    "#00ff80", "#0080ff", "#8000ff", "#ff0080", "#ff7373", "#73ff73", "#7373ff", "#ffff73",
    "#ff73ff", "#73ffff", "#ffb973", "#b9ff73", "#73ffb9", "#73b9ff", "#b973ff", "#ff73b9",
    "#b80000", "#00b800", "#4949b8", "#b8b800", "#b800b8", "#00b8b8", "#b85c00", "#5cb800",
    "#00b85c", "#005cb8", "#732eb8", "#b8005c", "#d94141", "#41d941", "#4141d9", "#d9d941",
    "#d941d9", "#41d9d9", "#d98d41", "#8dd941", "#41d98d", "#418dd9", "#8d41d9", "#d9418d",
    "#ffffff",
];

#[derive(Clone, Copy, Debug, Default, Serialize, Deserialize, PartialEq, Eq)]
#[serde(rename_all = "lowercase")]
pub enum Mode {
    #[default]
    #[serde(alias = "identifier", alias = "id", alias = "levels")]
    Level,
    #[serde(alias = "chips")]
    Chip,
    #[serde(alias = "layers")]
    Layer,
}
impl Mode {
    pub fn parse(text: &str) -> Result<Self> {
        match text.to_ascii_lowercase().as_str() {
            "level" | "identifier" | "id" | "levels" => Ok(Self::Level),
            "chip" | "chips" => Ok(Self::Chip),
            "layer" | "layers" => Ok(Self::Layer),
            _ => Err(Error::input("jobdeck mode must be level, chip or layer")),
        }
    }
}
#[derive(Clone, Debug, Serialize)]
pub struct ColorScheme {
    pub mode: Mode,
    pub palette: Vec<String>,
    pub overrides: BTreeMap<String, String>,
    pub cross_ly_dt: bool,
    pub fallback: String,
}
impl Default for ColorScheme {
    fn default() -> Self {
        Self {
            mode: Mode::Level,
            palette: PALETTE.into_iter().map(str::to_owned).collect(),
            overrides: BTreeMap::new(),
            cross_ly_dt: true,
            fallback: "#808080".into(),
        }
    }
}
#[derive(Deserialize)]
#[serde(untagged)]
enum PaletteSpec {
    Named(String),
    Colors(Vec<String>),
}
#[derive(Deserialize, Default)]
#[serde(default, deny_unknown_fields)]
struct SavedScheme {
    mode: Option<String>,
    palette: Option<PaletteSpec>,
    overrides: Option<BTreeMap<String, String>>,
    cross_ly_dt: Option<bool>,
    fallback: Option<String>,
}
#[derive(Clone, Debug, PartialEq, Eq, PartialOrd, Ord)]
pub enum ColorKey {
    Level(i64),
    Chip(String),
    Layer(i64, i64),
}
impl ColorKey {
    fn text(&self) -> String {
        match self {
            Self::Level(i) => i.to_string(),
            Self::Chip(c) => c.clone(),
            Self::Layer(l, d) => format!("{l}/{d}"),
        }
    }
}
#[derive(Clone, Debug, Serialize)]
pub struct OrderRow {
    pub pos: usize,
    pub key: String,
    pub color: String,
    pub name: String,
    pub source: String,
}
#[derive(Debug)]
pub struct ColorMap {
    pub values: BTreeMap<ColorKey, String>,
    pub order: Vec<OrderRow>,
}
impl ColorMap {
    pub fn report(&self, mode: Mode) -> Value {
        let map: BTreeMap<_, _> = self.values.iter().map(|(k, v)| (k.text(), v)).collect();
        json!({"mode":mode,"order":self.order,"map":map})
    }
}
impl ColorScheme {
    pub fn load(path: &Path) -> Result<Self> {
        let f = regular_file(path)?;
        let mut bytes = Vec::new();
        f.take(4 * 1024 * 1024 + 1).read_to_end(&mut bytes)?;
        if bytes.len() > 4 * 1024 * 1024 {
            return Err(Error::input("color scheme exceeds 4 MiB"));
        }
        Self::from_json(&bytes)
    }
    pub fn from_json(bytes: &[u8]) -> Result<Self> {
        let saved: SavedScheme = serde_json::from_slice(bytes)
            .map_err(|e| Error::input(format!("invalid color scheme: {e}")))?;
        let mut scheme = Self::default();
        if let Some(mode) = saved.mode {
            scheme.mode = Mode::parse(&mode)?;
        }
        if let Some(palette) = saved.palette {
            scheme.palette = match palette {
                PaletteSpec::Named(name) => match name.to_ascii_lowercase().as_str() {
                    "jobdeck" => PALETTE.to_vec(),
                    "reserve" => RESERVE.to_vec(),
                    _ => {
                        return Err(Error::input(
                            "unknown palette; use jobdeck, reserve, or a list",
                        ))
                    }
                }
                .into_iter()
                .map(str::to_owned)
                .collect(),
                PaletteSpec::Colors(colors) => colors,
            };
        }
        if let Some(overrides) = saved.overrides {
            scheme.overrides = overrides;
        }
        if let Some(cross) = saved.cross_ly_dt {
            scheme.cross_ly_dt = cross;
        }
        if let Some(fallback) = saved.fallback {
            scheme.fallback = fallback;
        }
        scheme.validate()?;
        Ok(scheme)
    }
    pub fn validate(&self) -> Result<()> {
        if self.palette.is_empty() || self.palette.len() > 65_536 {
            return Err(Error::input("color palette must have 1..65536 entries"));
        }
        for color in self
            .palette
            .iter()
            .chain(self.overrides.values())
            .chain(std::iter::once(&self.fallback))
        {
            if color.len() > 128
                || !color.is_ascii()
                || color.chars().any(|c| c.is_whitespace() || c.is_control())
            {
                return Err(Error::input("invalid color token"));
            }
        }
        Ok(())
    }
    pub fn palette_color(&self, position: usize) -> &str {
        // Public model fields can be edited before validate/build. No panic
        // on a temporarily empty palette; build/view still reject it.
        if self.palette.is_empty() {
            &self.fallback
        } else {
            &self.palette[position % self.palette.len()]
        }
    }
    fn pinned(&self, key: &str) -> Option<&str> {
        let bare = key
            .strip_prefix("CHIP:")
            .or_else(|| key.strip_prefix('$'))
            .or_else(|| key.strip_prefix("LY"))?;
        self.overrides
            .get(bare)
            .filter(|s| !s.is_empty())
            .map(String::as_str)
    }
    pub fn build(&self, deck: &JobDeck, ids: Option<&BTreeSet<i64>>) -> Result<ColorMap> {
        self.validate()?;
        let chip_ids = chip_ids(deck);
        let pairs = if self.mode == Mode::Layer {
            layer_pairs(deck, self.cross_ly_dt)?
        } else {
            BTreeSet::new()
        };
        let order: Vec<String> = match self.mode {
            Mode::Level => deck.levels().iter().map(|i| format!("${i}")).collect(),
            Mode::Chip => chip_order(&chip_ids, ids.unwrap_or(&deck.levels())),
            Mode::Layer => pairs
                .iter()
                .map(|(l, _)| *l)
                .collect::<BTreeSet<_>>()
                .iter()
                .map(|l| format!("LY{l}"))
                .collect(),
        };
        let order: Vec<_> = order
            .into_iter()
            .enumerate()
            .map(|(pos, key)| {
                let pinned = self.pinned(&key);
                let color = pinned.unwrap_or_else(|| self.palette_color(pos)).to_owned();
                OrderRow {
                    pos,
                    key,
                    name: color_name(&color).into(),
                    color,
                    source: if pinned.is_some() {
                        "pinned"
                    } else {
                        "palette"
                    }
                    .into(),
                }
            })
            .collect();
        let lookup: BTreeMap<_, _> = order
            .iter()
            .map(|r| (r.key.as_str(), r.color.as_str()))
            .collect();
        let mut values = BTreeMap::new();
        match self.mode {
            Mode::Level => {
                for level in deck.levels() {
                    values.insert(
                        ColorKey::Level(level),
                        lookup[format!("${level}").as_str()].into(),
                    );
                }
            }
            Mode::Chip => {
                for chip in chip_ids {
                    values.insert(
                        ColorKey::Chip(chip.into()),
                        lookup[format!("CHIP:{chip}").as_str()].into(),
                    );
                }
            }
            Mode::Layer => {
                for (ly, dt) in pairs {
                    let color = self
                        .overrides
                        .get(&format!("{ly}/{dt}"))
                        .filter(|s| !s.is_empty())
                        .map(String::as_str)
                        .unwrap_or(lookup[format!("LY{ly}").as_str()]);
                    values.insert(ColorKey::Layer(ly, dt), color.into());
                }
            }
        }
        Ok(ColorMap { values, order })
    }
    pub fn color_of<'a>(&'a self, map: &'a ColorMap, placement: &Placement) -> &'a str {
        let key = match self.mode {
            Mode::Level => ColorKey::Level(placement.idx),
            Mode::Chip => ColorKey::Chip(placement.chip.clone()),
            Mode::Layer => ColorKey::Layer(placement.ly, placement.dt),
        };
        map.values.get(&key).map_or(&self.fallback, String::as_str)
    }
}
pub fn chip_ids(deck: &JobDeck) -> Vec<&str> {
    let mut seen = BTreeSet::new();
    deck.chips
        .iter()
        .filter_map(|c| seen.insert(c.id.as_str()).then_some(c.id.as_str()))
        .collect()
}
pub fn layer_pairs(deck: &JobDeck, cross: bool) -> Result<BTreeSet<(i64, i64)>> {
    let mut pairs = BTreeSet::new();
    let mut visited = 0usize;
    for e in deck.chips.iter().flat_map(|c| &c.entries) {
        let row = entry_pairs(e, cross, 65_536)?;
        visited += row.len();
        if visited > 4_000_000 {
            return Err(Error::input("jobdeck color table visit limit exceeded"));
        }
        pairs.extend(row);
        if pairs.len() > 65_536 {
            return Err(Error::input("jobdeck color table size limit exceeded"));
        }
    }
    Ok(pairs)
}
pub fn chip_order(chips: &[&str], ids: &BTreeSet<i64>) -> Vec<String> {
    let mut out = Vec::new();
    let mut taken = 0;
    for &level in ids {
        let want = level.saturating_sub(1).max(0).min(chips.len() as i64) as usize;
        while taken < want {
            out.push(format!("CHIP:{}", chips[taken]));
            taken += 1;
        }
        out.push(format!("${level}"));
    }
    out.extend(chips[taken..].iter().map(|c| format!("CHIP:{c}")));
    out
}
fn color_name(color: &str) -> &str {
    PALETTE
        .iter()
        .position(|p| p.eq_ignore_ascii_case(color))
        .map_or(color, |i| NAMES[i])
}

#[cfg(test)]
mod tests {
    use super::*;
    #[test]
    fn order_pins_and_aliases() {
        assert_eq!(
            chip_order(&["C1", "C2", "C3"], &BTreeSet::from([1, 3])),
            ["$1", "CHIP:C1", "CHIP:C2", "$3", "CHIP:C3"]
        );
        assert_eq!(Mode::parse("IDENTIFIER").unwrap(), Mode::Level);
        assert!(ColorScheme::from_json(br#"{"palette":[]}"#).is_err());
        assert!(ColorScheme::from_json(br#"{"palette":["red\nsource"]}"#).is_err());
        let scheme = ColorScheme::from_json(
            br##"{"mode":"chip","palette":"reserve","overrides":{"C1":"#123456"}}"##,
        )
        .unwrap();
        assert_eq!(scheme.palette.len(), 49);
        assert_eq!(scheme.pinned("CHIP:C1"), Some("#123456"));
    }
}
